File: app/src/main/java/com/example/vloboda/dynamicentityeditor/dynamic/EditTextFactory.java
package com.example.vloboda.dynamicentityeditor.dynamic;

import android.content.Context;
import android.support.design.widget.TextInputLayout;
import android.text.InputType;
import android.view.LayoutInflater;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;

import com.example.bl.ui.DynamicEditableViewBase;
import com.example.bl.ui.DynamicView;
import com.example.bl.ui.ViewFactoryBase;
import com.example.bl.ui.DynamicViewContainer;
import com.example.bl.ui.UICommonServices;
import com.example.model.FieldConfiguration;
import com.example.model.EditObjectTemplateDto;
import com.example.model.FieldDto;
import com.example.model.StringHelperKt;
import com.example.vloboda.dynamicentityeditor.R;


public class EditTextFactory extends ViewFactoryBase {

    protected LayoutInflater inflater;

    EditTextFactory(UICommonServices commonServices) {
        super(commonServices);
    }

    protected int getInputType() {
        return InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_FLAG_CAP_SENTENCES;
    }

    private View getMainView(LayoutInflater inflater) {
        return inflater.inflate(R.layout.dynamic_edit_text, null);
    }

    @Override
    public DynamicView getView(DynamicViewContainer container,
                               final FieldConfiguration configuration) {
        inflater = (LayoutInflater)container.getLayoutInflater();

        View mainView = getMainView(inflater);
        final TextInputLayout textLayout = mainView.findViewById(R.id.det_text_layout);
        textLayout.setHintAnimationEnabled(false);
        final EditText editText = mainView.findViewById(R.id.det_main_text);
        if (configuration.hint != null) {
            textLayout.setHint(configuration.hint);
        } else {
            textLayout.setHint(configuration.caption);
        }
        editText.setInputType(getInputType());
        editText.setTag(configuration.fieldCode);

        editText.setOnEditorActionListener((textView, actionId, keyEvent) -> {
            if(actionId == EditorInfo.IME_ACTION_GO){
                editText.clearFocus();
                InputMethodManager imm = (InputMethodManager)inflater.getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
                if (imm != null) {
                    imm.hideSoftInputFromWindow(editText.getWindowToken(), 0);
                }
            }
            return false;
        });
        textLayout.setHintAnimationEnabled(true);

        final DynamicEditableViewBase dynamicView = new DynamicEditableViewBase(getLocationManager()) {

            @Override
            public View getView() {
                return mainView;
            }

            @Override
            protected FieldDto getCurrentValue() {
                return new FieldDto(configuration.fieldCode, editText.getText().toString());
            }

            @Override
            public void setValue(final FieldDto attribute) {
                setInitialValue(attribute);
                textLayout.setHintAnimationEnabled(false);
                editText.setText(attribute.getValue());
                textLayout.setHintAnimationEnabled(true);

                editText.setOnLongClickListener(v -> {
                    if (attribute.getChangeInfo() != null) {
                        getToaster().showLongMessage(attribute.getChangeInfo().toString());
                    }
                    return true;
                });
            }

            @Override
            public String getFieldCode() {
                return configuration.fieldCode;
            }

            @Override
            public void setTemplate(EditObjectTemplateDto.ItemDto template) {
                if (!StringHelperKt.isNullOrEmpty(template.getCaption())) {
                    textLayout.setHint(template.getCaption());
                }
            }
        };

        editText.setOnFocusChangeListener((v, hasFocus) -> {
            if (!hasFocus) {
                dynamicView.rememberChangeLocation();
            }
        });

        return dynamicView;
    }
}

File: app/src/main/java/com/example/vloboda/dynamicentityeditor/dynamic/EditTextFactory.kt
package com.example.vloboda.dynamicentityeditor.dynamic

import android.content.Context
import android.support.design.widget.TextInputLayout
import android.text.InputType
import android.view.LayoutInflater
import android.view.View
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.EditText

import com.example.bl.ui.DynamicEditableViewBase
import com.example.bl.ui.DynamicView
import com.example.bl.ui.ViewFactoryBase
import com.example.bl.ui.DynamicViewContainer
import com.example.bl.ui.UICommonServices
import com.example.model.FieldConfiguration
import com.example.model.EditObjectTemplateDto
import com.example.model.FieldDto
import com.example.model.*
import com.example.vloboda.dynamicentityeditor.R

open class EditTextFactory internal constructor(commonServices: UICommonServices) : ViewFactoryBase(commonServices) {

    protected lateinit var inflater: LayoutInflater

    protected open val inputType: Int
        get() = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_FLAG_CAP_SENTENCES

    private fun getMainView(inflater: LayoutInflater): View {
        return inflater.inflate(R.layout.dynamic_edit_text, null)
    }

    override fun getView(container: DynamicViewContainer,
                         configuration: FieldConfiguration): DynamicView {
        inflater = container.layoutInflater as LayoutInflater

        val mainView = getMainView(inflater)
        val textLayout = mainView.findViewById<TextInputLayout>(R.id.det_text_layout)
        textLayout.isHintAnimationEnabled = false
        val editText = mainView.findViewById<EditText>(R.id.det_main_text)
        if (configuration.hint != null) {
            textLayout.hint = configuration.hint
        } else {
            textLayout.hint = configuration.caption
        }
        editText.inputType = inputType
        editText.tag = configuration.fieldCode

        editText.setOnEditorActionListener { textView, actionId, keyEvent ->
            if (actionId == EditorInfo.IME_ACTION_GO) {
                editText.clearFocus()
                val imm = inflater.context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                imm.hideSoftInputFromWindow(editText.windowToken, 0)
            }
            false
        }
        textLayout.isHintAnimationEnabled = true

        val dynamicView = object : DynamicEditableViewBase(locationManager) {

            override val view: View
                get() = mainView

            override val currentValue: FieldDto
                get() = FieldDto(configuration.fieldCode, editText.text.toString())

            override val fieldCode: String
                get() = configuration.fieldCode

            override fun setValue(value: FieldDto) {
                initialValue = value
                textLayout.isHintAnimationEnabled = false
                editText.setText(value.value)
                textLayout.isHintAnimationEnabled = true

                editText.setOnLongClickListener {
                    if (value.changeInfo != null) {
                        toaster.showLongMessage(value.changeInfo.toString())
                    }
                    true
                }
            }

            override fun setTemplate(templateItem: EditObjectTemplateDto.ItemDto) {
                if (!templateItem.Caption.isNullOrEmpty()) {
                    textLayout.hint = templateItem.Caption
                }
            }
        }

        editText.setOnFocusChangeListener { v, hasFocus ->
            if (!hasFocus) {
                dynamicView.rememberChangeLocation()
            }
        }

        return dynamicView
    }
}

